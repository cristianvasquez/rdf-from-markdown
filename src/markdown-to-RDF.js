import { simpleAst } from 'docs-and-graphs'
import { astTriplifier } from './triplifiers/astTriplifier.js'

const defaultOptions = {
  splitOnTag: false,
  splitOnHeader: false,
  splitOnId: true,
  addLabels: false,
  includeWikipaths: true,
}

function markdownToRDF (fullText, { termMapper, pointer, path }, options = {}) {

  const json = simpleAst(fullText, { normalize: true, inlineAsArray: true })
  return astTriplifier(json, {
    pointer, termMapper, path,
  }, { ...defaultOptions, ...options })

}

export { markdownToRDF }
