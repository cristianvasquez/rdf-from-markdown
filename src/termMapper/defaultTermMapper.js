import rdf from '../rdf-ext.js'
import { SimpleUriMinter } from './uriMinter.js'

function createTermMapper ({
  baseNamespace, customMapper, getPathByName,
}) {

  const pathUriMinter = new SimpleUriMinter(`${baseNamespace().value}note/`)
  const propertyUriMinter = new SimpleUriMinter(`${baseNamespace().value}property/`)

  const maybeMapped = (txt) => customMapper ? customMapper(txt) : undefined

  // returns: NamedNode
  function pathToUri (path) {
    return pathUriMinter.toUri(path)
  }

  // To build properties
  // For example, "has name" -> http://some-vault/has-name
  function newProperty (txt, options) {
    return propertyUriMinter.toUri(txt)
  }

  function newLiteral (txt, options) {
    return rdf.literal(txt)
  }

  // http://example.com/ + ^blockId -> http://example.com/blockId
  const blockUri = (uri, blockId) => rdf.namedNode(
    `${uri.value}/${blockId.replace(/^\^/, '')}`)

  return {
    newLiteral,
    newProperty,
    maybeMapped,
    blockUri,
    pathToUri,
    pathUriMinter,
    propertyUriMinter,
    getPathByName: getPathByName ? getPathByName.path : (x) => `${x}.md`,
  }

}

export {
  createTermMapper,
}
