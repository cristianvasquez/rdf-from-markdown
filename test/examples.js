import dataset from '@rdfjs/dataset'
import { readFile } from 'fs/promises'
import { resolve } from 'path'
import { createTriplifier } from '../index.js'
import ns from '../src/namespaces.js'
import { expect } from 'expect'
import { markdownToRDF } from '../src/markdown-to-RDF.js'
import toMatchSnapshot from 'expect-mocha-snapshot'
import rdf from '../src/rdf-ext.js'
import { createTermMapper } from '../src/termMapper/defaultTermMapper.js'
import { prettyPrint } from './support/serialization.js'
import {
  allTests, splitOnTags, splitOnIdentifiers, splitOnHeaders, yamlLike,
} from './tests.js'

expect.extend({ toMatchSnapshot })

// Optional function to map strings to URIs
const customMapper = (str, context) => {
  // It's of the form schema::name
  if (str.split(':').length === 2) {
    const [vocabulary, property] = str.split(':')
    return ns[vocabulary] ? ns[vocabulary][property] : undefined
  }

  const values = {
    'is a': ns.rdf.type,
  }
  return values [str]
}

const dir = 'test/markdown'
const triplifier = await createTriplifier(dir, {
  baseNamespace: ns.ex, customMapper,
})

describe('triplify examples', async function () {

  for (const file of triplifier.vault.getFiles()) {
    it(file, async function () {

      const text = await readFile(resolve(dir, file), 'utf8')
      const pointer = triplifier.toRDF(text, { path: file },
        { addLabels: true, includeWikiPaths: true, splitOnHeader: true })

      const pretty = await prettyPrint(pointer.dataset)
      expect(pretty).toMatchSnapshot(this)
    })
  }
})


