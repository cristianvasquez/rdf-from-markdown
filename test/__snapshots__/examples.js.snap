// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`triplify examples Alice.md 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/QWxpY2UubWQ=> schema:name \\"Alice\\" ;
	dot:contains [
		schema:name \\"Alice\\" ;
		dot:linktext \\"Alice.md#Alice\\" ;
		<http://example.org/property/aXMgYQ==> <http://example.org/note/UGVyc29u> ;
		dot:related <http://example.org/note/SWNlIGNyZWFt> ;
	] .

<http://example.org/note/UGVyc29u> schema:name \\"Person\\" ;
	dot:wikipath \\"Person\\" .

<http://example.org/property/aXMgYQ==> schema:name \\"is a\\" .

<http://example.org/note/SWNlIGNyZWFt> schema:name \\"Ice cream\\" .

"
`;

exports[`triplify examples Person.md 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/UGVyc29uLm1k> schema:name \\"Person\\" ;
	dot:contains [
		schema:name \\"Person\\" ;
		dot:linktext \\"Person.md#Person\\" ;
		dot:contains [
			schema:name \\"Section\\" ;
			dot:linktext \\"Person.md#Section\\" ;
			<http://example.org/property/cHJvcGVydHk=> \\"A\\" ;
		], [
			schema:name \\"Subsection\\" ;
			dot:linktext \\"Person.md#Subsection\\" ;
			<http://example.org/property/cHJvcGVydHk=> \\"C\\" ;
		], [
			schema:name \\"Subsection\\" ;
			dot:linktext \\"Person.md#Subsection\\" ;
			<http://example.org/property/cHJvcGVydHk=> \\"B\\" ;
			dot:tag \\"person\\" ;
		] ;
		<http://example.org/property/ZGVzY3JpcHRpb24=> \\"the class of persons\\" ;
	] .

<http://example.org/property/ZGVzY3JpcHRpb24=> schema:name \\"description\\" .

<http://example.org/property/cHJvcGVydHk=> schema:name \\"property\\" .

"
`;

exports[`triplify examples Test Canvas.canvas 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/QWxpY2UubWQ=> schema:name \\"Alice\\" ;
	dot:wikipath \\"Alice.md\\" .

<http://example.org/note/UGVyc29uLm1k> schema:name \\"Person\\" ;
	dot:wikipath \\"Person.md\\" .

<http://example.org/note/Ym9iL0JvYiBEZXRhaWxzLm1k> schema:name \\"Bob Details\\" ;
	dot:wikipath \\"bob/Bob Details.md\\" .

<http://example.org/note/Ym9iL0JvYi5tZA==> schema:name \\"Bob\\" ;
	dot:wikipath \\"bob/Bob.md\\" .

<http://example.org/note/aG91c2VzL0JvYkhvdXNlLm1k> schema:name \\"BobHouse\\" ;
	dot:wikipath \\"houses/BobHouse.md\\" .

<http://example.org/note/aG91c2VzL2ltZy5wbmc=> schema:name \\"img.png\\" ;
	dot:wikipath \\"houses/img.png\\" .

<http://example.org/note/VGVzdCBDYW52YXMuY2FudmFz> schema:name \\"Test Canvas.canvas\\" ;
	dot:contains [
		schema:name \\"Entities\\" ;
		dot:contains [
			schema:name \\"People\\" ;
			dot:contains <http://example.org/note/QWxpY2UubWQ=>, <http://example.org/note/Ym9iL0JvYi5tZA==> ;
		], <http://example.org/note/QWxpY2UubWQ=>, <http://example.org/note/UGVyc29uLm1k>, <http://example.org/note/Ym9iL0JvYiBEZXRhaWxzLm1k>, <http://example.org/note/Ym9iL0JvYi5tZA==> ;
	], <http://example.org/note/aG91c2VzL0JvYkhvdXNlLm1k>, [
		schema:name \\"Resources\\" ;
		dot:contains <http://example.org/note/aG91c2VzL2ltZy5wbmc=> ;
	] .

"
`;

exports[`triplify examples Yaml.md 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/WWFtbC5tZA==> dot:tag \\"cats\\", \\"dogs\\" ;
	<http://example.org/property/ZGF0ZSB1cGRhdGVk> \\"2021-08-05T17:10:34+02:00\\" ;
	<http://example.org/property/dG9ncw==> \\"cats\\", \\"dogs\\" ;
	<http://example.org/property/b25l> \\"pepe\\", \\"carlos\\" ;
	<http://example.org/property/YW4tdXJp> <http://some-uri.com> ;
	schema:name \\"Yaml\\" .

<http://example.org/property/ZGF0ZSB1cGRhdGVk> schema:name \\"date updated\\" .

<http://example.org/property/dG9ncw==> schema:name \\"togs\\" .

<http://example.org/property/b25l> schema:name \\"one\\" .

<http://example.org/property/YW4tdXJp> schema:name \\"an-uri\\" .

"
`;

exports[`triplify examples bob/Bob Details.md 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/Ym9iL0JvYiBEZXRhaWxzLm1k> schema:name \\"Bob Details\\" ;
	dot:contains [
		schema:name \\"More details about Bob\\" ;
		dot:linktext \\"bob/Bob Details.md#More details about Bob\\" ;
		dot:contains [
			schema:name \\"Section\\" ;
			dot:linktext \\"bob/Bob Details.md#Section\\" ;
			<http://example.org/property/ZGVzY3JpcHRpb24=> \\"Section 1\\" ;
			<http://example.org/property/c29tZXdoYXQgcmVsYXRlZA==> <http://example.org/note/Qm9iIERldGFpbHMjXjI=> ;
			<http://example.org/property/cG9pbnRzIHRvIHRoZSBkb2N1bWVudA==> <http://example.org/note/Qm9iIERldGFpbHMjQSBoZWFkZXI=> ;
		], [
			schema:name \\"Section\\" ;
			dot:linktext \\"bob/Bob Details.md#Section\\" ;
			<http://example.org/property/ZGVzY3JpcHRpb24=> \\"Section 2\\" ;
			dot:related <http://example.org/note/Qm9iIERldGFpbHMjXjMg> ;
		], [
			schema:name \\"Section\\" ;
			dot:linktext \\"bob/Bob Details.md#Section\\" ;
			<http://example.org/property/ZGVzY3JpcHRpb24=> \\"Section 3\\" ;
			dot:related <http://example.org/note/VW5rbm93biNeMQ==> ;
		] ;
	] .

<http://example.org/property/ZGVzY3JpcHRpb24=> schema:name \\"description\\" .

<http://example.org/note/Qm9iIERldGFpbHMjXjI=> schema:name \\"Bob Details#^2\\" ;
	dot:wikipath \\"Bob Details#^2\\" .

<http://example.org/property/c29tZXdoYXQgcmVsYXRlZA==> schema:name \\"somewhat related\\" .

<http://example.org/note/Qm9iIERldGFpbHMjQSBoZWFkZXI=> schema:name \\"Bob Details#A header\\" ;
	dot:wikipath \\"Bob Details#A header\\" .

<http://example.org/property/cG9pbnRzIHRvIHRoZSBkb2N1bWVudA==> schema:name \\"points to the document\\" .

<http://example.org/note/Qm9iIERldGFpbHMjXjMg> schema:name \\" Pointer to section 3\\" .

<http://example.org/note/VW5rbm93biNeMQ==> schema:name \\"Unknown#^1\\" .

"
`;

exports[`triplify examples bob/Bob.md 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/aW1nLnBuZw==> dot:wikipath \\"img.png\\" ;
	schema:name \\"img.png\\" .

<http://example.org/note/Ym9iL0JvYi5tZA==> <http://example.org/property/d2l0aCBpbWFnZQ==> <http://example.org/note/aW1nLnBuZw==> ;
	dot:related <http://example.org/note/TGluaw==>, <http://something.com> ;
	<http://example.org/property/bGl2ZXMgaW4=> <http://example.org/note/Ym9iL2hvdXNlcy9Cb2JIb3VzZS5tZA==> ;
	<http://example.org/property/aGFzIGRldGFpbHM=> \\"[Bob Details.md](./bob/Bob Details.md)\\" ;
	<http://example.org/property/ZXg6a25vd3M=> <http://example.org/note/Ym9iL0FsaWNlLm1k> ;
	<http://example.org/property/aGFzIGltYWdlIDE=> <http://example.org/note/Ym9iL2hvdXNlcy9pbWcucG5n> ;
	<http://example.org/property/aXMgYQ==> <http://example.org/note/UGVyc29uLm1k> ;
	schema:name \\"Bob\\" .

<http://example.org/property/d2l0aCBpbWFnZQ==> schema:name \\"with image\\" .

<http://example.org/note/TGluaw==> schema:name \\"Link\\" .

<http://example.org/note/Ym9iL2hvdXNlcy9Cb2JIb3VzZS5tZA==> dot:wikipath \\"bob/houses/BobHouse.md\\" ;
	schema:name \\"BobHouse\\" .

<http://example.org/property/bGl2ZXMgaW4=> schema:name \\"lives in\\" .

<http://example.org/property/aGFzIGRldGFpbHM=> schema:name \\"has details\\" .

<http://example.org/note/Ym9iL0FsaWNlLm1k> dot:wikipath \\"bob/Alice.md\\" ;
	schema:name \\"Alice\\" .

<http://example.org/property/ZXg6a25vd3M=> schema:name \\"ex:knows\\" .

<http://example.org/note/Ym9iL2hvdXNlcy9pbWcucG5n> dot:wikipath \\"bob/houses/img.png\\" ;
	schema:name \\"img.png\\" .

<http://example.org/property/aGFzIGltYWdlIDE=> schema:name \\"has image 1\\" .

<http://example.org/note/UGVyc29uLm1k> dot:wikipath \\"Person.md\\" ;
	schema:name \\"Person\\" .

<http://example.org/property/aXMgYQ==> schema:name \\"is a\\" .

"
`;

exports[`triplify examples dot triples.md 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/ZG90IHRyaXBsZXMubWQ=> schema:name \\"dot triples\\" ;
	dot:contains [
		schema:name \\"Dot triples syntax\\" ;
		dot:linktext \\"dot triples.md#Dot triples syntax\\" ;
		dot:contains [
			schema:name \\"Alice\\" ;
			dot:linktext \\"dot triples.md#Alice\\" ;
			<http://example.org/property/aGFz> \\"37 years old\\" ;
			<http://example.org/property/ZXg6a25vd3M=> <http://example.org/note/Qm9i> ;
			dot:related <http://example.org/note/Qm9i>, <http://example.org/note/QWxpY2U=> ;
		], [
			schema:name \\"Embedded\\" ;
			dot:linktext \\"dot triples.md#Embedded\\" ;
			<http://example.org/property/RUE=> \\"EB\\" ;
			<http://example.org/property/RUM=> \\"ED\\" ;
		] ;
	] .

<http://example.org/property/aGFz> schema:name \\"has\\" .

<http://example.org/note/Qm9i> schema:name \\"Bob\\" ;
	dot:wikipath \\"Bob\\" .

<http://example.org/property/ZXg6a25vd3M=> schema:name \\"ex:knows\\" .

<http://example.org/note/QWxpY2U=> schema:name \\"Alice\\" .

<http://example.org/property/QQ==> <http://example.org/property/Qg==> \\"C\\" .

<http://example.org/property/Qg==> schema:name \\"B\\" .

<http://example.org/property/RUE=> schema:name \\"EA\\" .

<http://example.org/property/RUM=> schema:name \\"EC\\" .

"
`;

exports[`triplify examples houses/BobHouse.md 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/aG91c2VzL0JvYkhvdXNlLm1k> <http://example.org/property/aXMgYQ==> \\"House\\" ;
	schema:name \\"BobHouse\\" .

<http://example.org/property/aXMgYQ==> schema:name \\"is a\\" .

"
`;

exports[`triplify examples houses/img.png 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .
"
`;

exports[`triplify examples links.md 1`] = `
"@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix schema: <http://schema.org/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix ex: <http://example.org/> .
@prefix dot: <http://dottriples.org/> .
@prefix mark: <http://markdown.org/> .

<http://example.org/note/bGlua3MubWQ=> schema:name \\"links\\" ;
	dot:contains [
		schema:name \\"Links\\" ;
		dot:linktext \\"links.md#Links\\" ;
		dot:contains [
			schema:name \\"Produce two simple links\\" ;
			dot:linktext \\"links.md#Produce two simple links\\" ;
			dot:related <http://web2.org>, <http://web1.org> ;
		], [
			schema:name \\"With alias\\" ;
			dot:linktext \\"links.md#With alias\\" ;
			dot:related <https://www.rand.org/content/dam/rand/pubs/papers/2008/P899.pdf> ;
		] ;
	] .

<https://www.rand.org/content/dam/rand/pubs/papers/2008/P899.pdf> schema:name \\"1954\\" .

"
`;
